### PR TITLE
Implement push and pop for optimize

### PR DESCRIPTION
**Pull request by jklmnn**
ref #21

### CODE REVIEW

File: z3.ads [hunk 1]
```diff
@@ -402,6 +404,18 @@ package Z3 is  --  GCOV_EXCL_LINE
...
+
+   procedure Pop (Optimize : in out Z3.Optimize);
+
+   function Get_Number_Of_Values (Optimize : Z3.Optimize) return Natural;
+
+   procedure Get_Values (Optimize  :     Z3.Optimize;
+                         Constants : out Int_Array;
+                         Values    : out Int_Array) with
+      Pre => Constants'Length = Get_Number_Of_Values (Optimize)
+             and then Values'Length = Get_Number_Of_Values (Optimize);
```
Comment (senier) [inline z3.ads:417]: Shouldn't also `Value'Length = Constants'Length`?

**Review by senier — CHANGES_REQUESTED** [review]

File: z3.ads (hunk 1 above)
Comment (jklmnn) [reply z3.ads:417]: This is implied since `Get_Number_Of_Values` is not volatile.

File: z3.adb [hunk 2]
```diff
@@ -767,13 +775,76 @@ is
...
+
+   ------------------------------------------------------------------------------------------------
+
+   procedure Pop (Optimize : in out Z3.Optimize)
+   is
+      Objectives : constant Int_Maps.Map := Optimize.Objectives;
+   begin
+      if Optimize.Backtracking_Count < 1 then
+         raise Z3.Value_Error;
+      end if;
```
[the remark this answers is not in the document]
Comment (jklmnn) [reply z3.adb:803]: This property is not visible at the function spec so we can't use it in a precondition.

[the remark this answers is not in the document]
Comment (senier) [reply z3.adb:803]: Not this time ;)

**Review by senier — APPROVED** [review]

**Pull request merged**